// @livereader/graphly-d3/dist/component-vue3.mjs
import { defineComponent, ref, onMounted, watch, openBlock, createElementBlock, normalizeClass } from "vue";
import { ForceSimulation, Event } from "./main.mjs";
import "d3";
import "ajv";
const style = "";
const _hoisted_1 = {
  class: "gly-container",
  style: { "width": "100%", "height": "100%" }
};
const _sfc_main = /* @__PURE__ */ defineComponent({
  __name: "GraphlyD3",
  props: {
    dark: {
      type: Boolean,
      default: false
    },
    remoteOrigin: {
      type: String,
      default: ""
    },
    selectedNodes: {
      type: Array,
      default: () => []
    },
    envGravity: {
      type: Number,
      default: -1e4
    },
    linkDistance: {
      type: Number,
      default: 400
    },
    animationDuration: {
      type: Number,
      default: 300
    },
    draggableNodes: {
      type: Boolean,
      default: true
    },
    zoomEnabled: {
      type: Boolean,
      default: true
    },
    zoomScaleExtent: {
      type: Array,
      default: () => [0.1, 3]
    }
  },
  emits: [
    "node-click",
    "node-double-click",
    "node-context-menu",
    "node-drag-start",
    "node-drag-move",
    "node-drag-end",
    "link-click",
    "link-double-click",
    "link-context-menu",
    "link-drag-start",
    "link-drag-move",
    "link-drag-end",
    "environment-click",
    "environment-double-click",
    "environment-context-menu",
    "environment-move",
    "theme-change",
    "simulation-tick",
    "simulation-tick-end"
  ],
  setup(__props, { expose, emit }) {
    const props = __props;
    let svg = ref();
    let simulation = ref(null);
    expose({
      simulation
    });
    onMounted(() => {
      if (!svg.value)
        return;
      simulation.value = new ForceSimulation(svg.value);
      simulation.value.templateStore.remoteOrigin = props.remoteOrigin;
      simulation.value.selectedNodes = props.selectedNodes;
      simulation.value.envGravity = props.envGravity;
      simulation.value.linkDistance = props.linkDistance;
      simulation.value.animationDuration = props.animationDuration;
      simulation.value.draggableNodes = props.draggableNodes;
      simulation.value.zoomEnabled = props.zoomEnabled;
      simulation.value.zoomScaleExtent = [props.zoomScaleExtent[0], props.zoomScaleExtent[1]];
      simulation.value.on(Event.NodeClick, (e, node) => emit("node-click", e, node));
      simulation.value.on(Event.NodeDoubleClick, (e, node) => emit("node-double-click", e, node));
      simulation.value.on(Event.NodeContextMenu, (e, node) => emit("node-context-menu", e, node));
      simulation.value.on(
        Event.NodeDragStart,
        (e, node, position) => emit("node-drag-start", e, node, position)
      );
      simulation.value.on(
        Event.NodeDragMove,
        (e, node, position) => emit("node-drag-move", e, node, position)
      );
      simulation.value.on(
        Event.NodeDragEnd,
        (e, node, position) => emit("node-drag-end", e, node, position)
      );
      simulation.value.on(Event.LinkClick, (e, link) => emit("link-click", e, link));
      simulation.value.on(Event.LinkDoubleClick, (e, link) => emit("link-double-click", e, link));
      simulation.value.on(Event.LinkContextMenu, (e, link) => emit("link-context-menu", e, link));
      simulation.value.on(
        Event.LinkDragStart,
        (e, sourceNode, position) => emit("link-drag-start", e, sourceNode, position)
      );
      simulation.value.on(
        Event.LinkDragMove,
        (e, sourceNode, position) => emit("link-drag-move", e, sourceNode, position)
      );
      simulation.value.on(
        Event.LinkDragEnd,
        (e, sourceNode, targetNode, position) => emit("link-drag-end", e, sourceNode, targetNode, position)
      );
      simulation.value.on(
        Event.EnvironmentClick,
        (e, position) => emit("environment-click", e, position)
      );
      simulation.value.on(
        Event.EnvironmentDoubleClick,
        (e, position) => emit("environment-double-click", e, position)
      );
      simulation.value.on(
        Event.EnvironmentContextMenu,
        (e, position) => emit("environment-context-menu", e, position)
      );
      simulation.value.on(
        Event.EnvironmentMove,
        (transform) => emit("environment-move", transform)
      );
      simulation.value.on(Event.ThemeChange, (theme) => emit("theme-change", theme));
      simulation.value.on(Event.SimulationTick, () => emit("simulation-tick"));
      simulation.value.on(Event.SimulationTickEnd, () => emit("simulation-tick-end"));
    });
    watch(
      () => props.remoteOrigin,
      () => {
        if (!simulation.value)
          return;
        simulation.value.templateStore.remoteOrigin = props.remoteOrigin;
      }
    );
    watch(
      () => props.selectedNodes,
      () => {
        if (!simulation.value)
          return;
        simulation.value.selectedNodes = props.selectedNodes;
      }
    );
    watch(
      () => props.envGravity,
      () => {
        if (!simulation.value)
          return;
        simulation.value.envGravity = props.envGravity;
      }
    );
    watch(
      () => props.linkDistance,
      () => {
        if (!simulation.value)
          return;
        simulation.value.linkDistance = props.linkDistance;
      }
    );
    watch(
      () => props.animationDuration,
      () => {
        if (!simulation.value)
          return;
        simulation.value.animationDuration = props.animationDuration;
      }
    );
    watch(
      () => props.draggableNodes,
      () => {
        if (!simulation.value)
          return;
        simulation.value.draggableNodes = props.draggableNodes;
      }
    );
    watch(
      () => props.zoomEnabled,
      () => {
        if (!simulation.value)
          return;
        simulation.value.zoomEnabled = props.zoomEnabled;
      }
    );
    watch(
      () => props.zoomScaleExtent,
      () => {
        if (!simulation.value)
          return;
        simulation.value.zoomScaleExtent = [props.zoomScaleExtent[0], props.zoomScaleExtent[1]];
      }
    );
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock("div", _hoisted_1, [
        (openBlock(), createElementBlock("svg", {
          ref_key: "svg",
          ref: svg,
          style: { "width": "100%", "height": "100%" },
          class: normalizeClass({ dark: props.dark })
        }, null, 2))
      ]);
    };
  }
});
export {
  _sfc_main as default
};
